[MM-10569] Invite links won't work in mobile browser when the user is asked if they want to download the app

mickmister opened the pull request:
#### Summary
Fixes issue with first-time mobile viewers being indirected to sign up page incorrectly.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/8782
https://mattermost.atlassian.net/browse/MM-10569


#### Checklist

- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed

As @esethna pointed out in the github ticket, the query parameter for the invite link is dropped off when the user is redirected to the "Download App" page.
The `/` route is used for the "continue with browser" link, which in turn resolves to `/login`, so the user loses the team invite link's context.

For the solution, we are using a redirect url param to hook into the previous page.

esethna commented:
Thanks @mickmister, looks like this needs a rebase then we can test :)

mickmister commented:
PR is up to date

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0f0c69c944769644f.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0f0c69c944769644f

mattermod commented:
Spinmint test server destroyed

mickmister commented:
PR is up to date. I have changed the link to use `goBack()` so it can be agnostic to where the user came from. I have also added tests to get the two mobile app download components under test.

sudheerDev commented on tests/components/get_android_app/get_android_app.test.jsx:
```diff
@@ -0,0 +1,50 @@
+        link.simulate('click');
+        expect(goBack).toHaveBeenCalled();
+    });
+});
```
👍

lieut-data commented:
@esethna, @mickmister: running into unexpected behaviour on the spinmint above:

* login as `sysadmin`
* browse to https://i-0da12aeba34226193.test.spinmint.com/ad-1/channels/town-square
* delete the `__landingPageSeen__` bit from the local storage
* reload the page
* click `Continue with browser`

Instead of being sent back to town square, I'm sent to `/signup_user_complete` and then `/signup_email`, despite already being logged in. Can anyone else reproduce this?

I suspect it's because the spinmint is out of date -- I'll rebuild so we can reverify.

lieut-data commented on components/get_android_app/get_android_app.jsx:
```diff
@@ -59,12 +58,16 @@ export default function GetAndroidApp({androidAppDownloadLink}) {
                         link: (
-                            <Link to='/'>
+                            <a
+                                onClick={history.goBack}
```
I'm hesitant to rely on `history.goBack()` -- while it likely works now due to a bunch of redirects, if we ever allowed the `/get_ios_app` to be accessed directly (e.g. from a link outside the server), going back would be the wrong action to take.

I think it would generally be more robust if we modified the redirection to code in `root.jsx` to add a `redirect_to` querystring parameter to which the user should redirect afterwards, and have it default to `/` here when not provided for any reason.

mickmister replied:
For sure, redirect approach is much better. Using `history.goBack()` was giving off a strong smell.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0009d71c52b3fc46f.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0009d71c52b3fc46f

lieut-data commented:
Ok, looks like my previous report was specific to the old code. @esethna, given that the implementation has changed, I thought it might be good to have a second look again.

Unrelated to this change, it seems like the mobile app landing pages don't get triggered on `master` if you're already logged in -- rather, you are redirected away from there before the page even renders. I'll file a separate issue.

@mickmister, open to discussion on whether to proceed with `goBack` or take the potentially more robust `redirect_to` approach (or something similar).

sudheerDev commented:
> Unrelated to this change, it seems like the mobile app landing pages don't get triggered on master if you're already logged in -- rather, you are redirected away from there before the page even renders. I'll file a separate issue.

This is not related to this PR and have a fix for master https://github.com/mattermost/mattermost-webapp/pull/1963

mickmister commented on components/get_android_app/get_android_app.jsx:
```diff
@@ -59,12 +58,16 @@ export default function GetAndroidApp({androidAppDownloadLink}) {
                     values={{
                         link: (
-                            <Link to='/'>
+                            <a
```
Here's a comparison of the current link and a view of a button in its place. Which do you think looks most appropriate?

![image](https://user-images.githubusercontent.com/6913320/47698187-b295fa80-dbe4-11e8-9a3d-1f7803b5b3e7.png)
---
![image](https://user-images.githubusercontent.com/6913320/47698148-96925900-dbe4-11e8-8bb8-b16bcbd496cb.png)
---
![image](https://user-images.githubusercontent.com/6913320/47698362-72834780-dbe5-11e8-86e9-5f94a6a6cec3.png)

mickmister replied:
Got it, I figured that may have been what you meant. I can apply `role="button"` but I wasn't sure if that's what you were implying. I'll keep it as the anchor tag. Thanks for your feedback!

mickmister commented:
@lieut-data requested changes have been included

lieut-data commented:
Thanks, @mickmister -- I'll take another pass :)

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mickmister commented:
I've introduced some code duplication with the redirection logic. I'm thinking it could be extracted out into a general purpose utility function called `redirectOrRoot()`. What do you think?

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

lieut-data commented on components/get_android_app/get_android_app.jsx:
```diff
@@ -3,14 +3,24 @@
 
-export default function GetAndroidApp({androidAppDownloadLink}) {
+export default function GetAndroidApp({androidAppDownloadLink, history, location}) {
+    const clickedContinueWithBrowser = (e) => {
```
nit: `clickedContinueWithBrowser` seems like an awkward name for a handler. Perhaps `onContinue`?

lieut-data commented:
@mickmister, I'd probably skip the `redirectOrRoot` refactor. Technically it's duplicated, but it's not a lot of code and it only occurs twice at present.

lieut-data commented:
@mickmister, lastly can I get you to rebase this PR? Otherwise, it's good to merge :)

sudheerDev commented:
@lieut-data Feel free to merge if you are good with changes

lieut-data commented:
Thanks, @mickmister :)

mattermod commented:
Spinmint test server destroyed
